web/packages/teleport/src/Navigation/RecentHistory.tsx

```diff
@@ -0,0 +1,298 @@
+/**
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useState, useEffect, useRef } from 'react';
+import styled from 'styled-components';
+import { matchPath } from 'react-router';
+import { NavLink } from 'react-router-dom';
+import { Flex, Text, H4, P3, ButtonIcon } from 'design';
+import { Cross } from 'design/Icon';
+
+import { TeleportFeature } from 'teleport/types';
+import { useFeatures } from 'teleport/FeaturesContext';
+
+import { getSubsectionStyles } from './SideNavigation/Section';
+import { SidenavCategory } from './SideNavigation/categories';
+
+export type RecentHistoryItem = {
+  category?: SidenavCategory;
+  title: string;
+  route: string;
+  exact?: boolean;
+};
+
+type AnimatedItem = RecentHistoryItem & {
+  animationState: 'exiting' | 'entering' | '';
+};
+
+function getIconForRoute(
+  features: TeleportFeature[],
+  route: string
+): (props) => JSX.Element {
+  const feature = features.find(feature =>
+    matchPath(route, {
+      path: feature?.route?.path,
+      exact: false,
+    })
+  );
+
+  const icon = feature?.navigationItem?.icon || feature?.topMenuItem?.icon;
+  if (!icon) {
+    return () => null;
+  }
+
+  return icon;
+}
+
+export function RecentHistory({
+  recentHistoryItems,
+  onRemoveItem,
+}: {
+  recentHistoryItems: RecentHistoryItem[];
+  onRemoveItem: (route: string) => void;
+}) {
+  const features = useFeatures();
+  const [animatedItems, setAnimatedItems] = useState<AnimatedItem[]>([]);
+  const prevItemsRef = useRef<RecentHistoryItem[]>([]);
+
+  useEffect(() => {
+    const prevItems = prevItemsRef.current;
+    let newAnimatedItems = recentHistoryItems.map(item => ({
+      ...item,
+      animationState: '',
+    })) as AnimatedItem[];
+
+    const isFirstItemDeleted =
+      recentHistoryItems.findIndex(
+        item => item.route === prevItems[0]?.route
+      ) === -1;
```

Comment:
does doing `-1` without calculating full length risk going index out of range?

Reply:
`findIndex` returns -1 if it _doesn't_ find an element that matches the predicate, so this is safe and intended. All this is essentially doing is checking whether the first item which was first in `prevItems` no longer exists in `recentHistoryItems`.

> The findIndex() method of [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array) instances returns the index of the first element in an array that satisfies the provided testing function. If no elements satisfy the testing function, -1 is returned.

Reply:
`.includes()` wouldn't have worked in this case because in the case of arrays of objects, `.includes()` doesn't do a deep comparison and only checks the object reference (not its value), `findIndex` lets you provide a predicate function so you can compare the values of the fields themselves.